rust/bridge/jni/src/logging.rs

```diff
@@ -0,0 +1,145 @@
+//
+// Copyright 2021 Signal Messenger, LLC.
+// SPDX-License-Identifier: AGPL-3.0-only
+//
+
+use jni::objects::{GlobalRef, JClass, JObject, JValue};
+use jni::sys::jint;
+use jni::{JNIEnv, JavaVM};
+
+// Keep this in sync with SignalProtocolLogger.java, as well as the list below.
+#[derive(Clone, Copy)]
+enum JavaLogLevel {
+    Verbose = 2,
+    Debug = 3,
+    Info = 4,
+    Warn = 5,
+    Error = 6,
+    Assert = 7,
+}
+
+impl From<log::Level> for JavaLogLevel {
+    fn from(level: log::Level) -> Self {
+        use log::Level::*;
+        match level {
+            Error => Self::Error,
+            Warn => Self::Warn,
+            Info => Self::Info,
+            Debug => Self::Debug,
+            Trace => Self::Verbose,
+        }
+    }
+}
+
+impl From<JavaLogLevel> for jint {
+    fn from(level: JavaLogLevel) -> Self {
+        level as jint
+    }
+}
+
+impl From<JavaLogLevel> for JValue<'_> {
+    fn from(level: JavaLogLevel) -> Self {
+        Self::Int(level.into())
+    }
+}
+
+impl From<JavaLogLevel> for log::Level {
+    fn from(level: JavaLogLevel) -> Self {
+        use JavaLogLevel::*;
+        match level {
+            Error | Assert => Self::Error,
+            Warn => Self::Warn,
+            Info => Self::Info,
+            Debug => Self::Debug,
+            Verbose => Self::Trace,
+        }
+    }
+}
+
+struct JniLogger {
+    vm: JavaVM,
+    logger_class: GlobalRef,
+}
+
+impl JniLogger {
+    fn log_impl(&self, record: &log::Record) -> jni::errors::Result<()> {
+        let env = self.vm.attach_current_thread()?;
+        let level: JavaLogLevel = record.level().into();
+        let message = format!(
+            "{}:{}: {}",
+            record.file().unwrap_or("<unknown>"),
+            record.line().unwrap_or(0),
+            record.args(),
+        );
+        let args: [JValue; 3] = [
+            level.into(),
+            env.new_string("libsignal-client")?.into(),
+            env.new_string(message)?.into(),
+        ];
+        let result = env.call_static_method(
+            &self.logger_class,
+            "log",
+            "(ILjava/lang/String;Ljava/lang/String;)V",
+            &args,
+        );
+
+        let throwable = env.exception_occurred()?;
+        if **throwable == *JObject::null() {
+            result?;
+        } else {
+            env.exception_clear()?;
+        }
+        Ok(())
+    }
+}
+
+impl log::Log for JniLogger {
+    fn enabled(&self, _metadata: &log::Metadata) -> bool {
+        true
+    }
+
+    fn log(&self, record: &log::Record) {
+        if self.log_impl(record).is_err() {
+            // Drop the error; it's not like we can log it!
+        }
+    }
+
+    fn flush(&self) {}
+}
+
+#[no_mangle]
+pub unsafe extern "C" fn Java_org_signal_client_internal_Native_Logger_1Initialize(
+    env: JNIEnv,
+    _class: JClass,
+    max_level: jint,
+    logger_class: JClass,
+) {
+    let logger = JniLogger {
```

Comment:
It felt superfluous to me since there's not really any abstraction here. I think it's also more obvious why the `expect` is acceptable when it's directly in an entry point. But those are my only reasons, can move if you aren't convinced. :-)

Reply:
I figured the expect would still be here, `JniLogger::new` returning a `Result` with `?` then this is becomes
`let logger = JniLogger::new(...).expect("...")`
which saves having to read one `expect`.

I guess my preference is, whenever initializing a struct is more than just copying/moving values, write a constructor fn. But it's just my preference, *and* this code is really only ever called once, so what you have here is fine.